## Movie review Prediction System

In [1]:
## Step 1 Import all libraries and Load the model
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing import sequence
from tensorflow.keras.models import load_model

In [2]:
## load the imdb dataset word index

word_index = imdb.get_word_index()
reverse_word_index = {value:key for key,value in word_index.items()}

In [3]:
# Load the pre-Trained model with ReLU activation
model = load_model('best_model.h5')
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding (Embedding)           │ (None, 500, 128)       │     1,280,000 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ simple_rnn (SimpleRNN)          │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,313,027 (5.01 MB)

 Trainable params: 1,313,025 (5.01 MB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 2 (12.00 B)

In [4]:
## Step 2 Helper Function 
## Function to decode reviews

OFFSET = 3
def decode_review(encoded_review):
    return ' '.join([reverse_word_index.get(i-3,'?') for i in encoded_review])

## Function to preprocess user input 

def preprocess_text(text):
    words = text.lower().split()
    encoded_review = [word_index.get(word,2) + OFFSET for word in words]
    padded_review = sequence.pad_sequences([encoded_review],maxlen=500)
    return padded_review

In [5]:
### prediction function 

def predict_sentiment(review):
    preprocessed_input = preprocess_text(review)
    prediction = model.predict(preprocessed_input)

    sentiment = 'Positive' if prediction[0][0]> 0.5 else 'Negative'
    return sentiment, prediction[0][0]

In [6]:
## Step 4: User Input and Prediction 
## Example review for prediction 
reviews = ["great acting", "boring plot", "i loved it", "waste of time","The movie was really bad"]
for review in reviews:
    sentiment, score = predict_sentiment(review)
    print(f"Review: {review}\nSentiment: {sentiment}, Score: {score:.4f}\n")


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 425ms/step
Review: great acting
Sentiment: Negative, Score: 0.2005

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 81ms/step
Review: boring plot
Sentiment: Negative, Score: 0.0529

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 82ms/step
Review: i loved it
Sentiment: Positive, Score: 0.7723

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 75ms/step
Review: waste of time
Sentiment: Negative, Score: 0.0885

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 95ms/step
Review: The movie was really bad
Sentiment: Negative, Score: 0.0349



In [14]:

review = "Awesome mindblowing movie"
sentiment, score = predict_sentiment(review)
print(f"Review: {review}\nSentiment: {sentiment}, Score: {score:.4f}\n")


InvalidArgumentError: Graph execution error:

Detected at node sequential_1/embedding_1/GatherV2 defined at (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main

  File "<frozen runpy>", line 88, in _run_code

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\ipykernel_launcher.py", line 18, in <module>

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\ipykernel\kernelapp.py", line 739, in start

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\tornado\platform\asyncio.py", line 205, in start

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\asyncio\base_events.py", line 618, in run_forever

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\asyncio\base_events.py", line 1951, in _run_once

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\asyncio\events.py", line 84, in _run

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\ipykernel\kernelbase.py", line 545, in dispatch_queue

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\ipykernel\kernelbase.py", line 534, in process_one

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\ipykernel\kernelbase.py", line 437, in dispatch_shell

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\ipykernel\ipkernel.py", line 362, in execute_request

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\ipykernel\kernelbase.py", line 778, in execute_request

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\ipykernel\ipkernel.py", line 449, in do_execute

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\ipykernel\zmqshell.py", line 549, in run_cell

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\IPython\core\interactiveshell.py", line 3077, in run_cell

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\IPython\core\interactiveshell.py", line 3132, in _run_cell

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\IPython\core\async_helpers.py", line 128, in _pseudo_sync_runner

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\IPython\core\interactiveshell.py", line 3336, in run_cell_async

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\IPython\core\interactiveshell.py", line 3519, in run_ast_nodes

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\IPython\core\interactiveshell.py", line 3579, in run_code

  File "C:\Users\prash\AppData\Local\Temp\ipykernel_10160\3357125385.py", line 5, in <module>

  File "C:\Users\prash\AppData\Local\Temp\ipykernel_10160\263984499.py", line 5, in predict_sentiment

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\utils\traceback_utils.py", line 117, in error_handler

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\backend\tensorflow\trainer.py", line 566, in predict

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\backend\tensorflow\trainer.py", line 260, in one_step_on_data_distributed

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\backend\tensorflow\trainer.py", line 250, in one_step_on_data

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\backend\tensorflow\trainer.py", line 105, in predict_step

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\utils\traceback_utils.py", line 117, in error_handler

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\layers\layer.py", line 941, in __call__

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\utils\traceback_utils.py", line 117, in error_handler

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\ops\operation.py", line 59, in __call__

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\utils\traceback_utils.py", line 156, in error_handler

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\models\sequential.py", line 220, in call

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\models\functional.py", line 183, in call

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\ops\function.py", line 206, in _run_through_graph

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\models\functional.py", line 644, in call

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\utils\traceback_utils.py", line 117, in error_handler

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\layers\layer.py", line 941, in __call__

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\utils\traceback_utils.py", line 117, in error_handler

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\ops\operation.py", line 59, in __call__

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\utils\traceback_utils.py", line 156, in error_handler

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\layers\core\embedding.py", line 150, in call

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\ops\numpy.py", line 5795, in take

  File "c:\Users\prash\OneDrive\Desktop\Certificate_Practice\venv\Lib\site-packages\keras\src\backend\tensorflow\numpy.py", line 2340, in take

indices[0,498] = 45951 is not in [0, 10000)
	 [[{{node sequential_1/embedding_1/GatherV2}}]] [Op:__inference_one_step_on_data_distributed_300]

#### The model works well for known and simple reviews, such as:

"i loved it" → Positive (0.7723)

"boring plot" → Negative (0.0529)

#### Wrongly Classified:
"great acting" → Predicted: Negative (0.2005) – should be Positive ❌

Despite being a clearly positive phrase, it's misclassified.

Possible Reason: Short length, lack of context, or under-representation in training data.

#### It correctly captures clear sentiments but fails when words are out of vocabulary:

"Awesome mindblowing movie" → ❌ Error

Error: indices[0,498] = 45951 is not in [0, 10000)

Cause: "awesome" or "mindblowing" not present in the IMDB word index (only top 10,000 words are included by default)

### Solution:
Handle OOV (Out-of-Vocabulary) words:

Use num_words=10000 when loading imdb.get_word_index() (already default)

Replace unknown words with a special OOV token (e.g., index 2) manually or using Tokenizer(oov_token='<OOV>') if using custom data

Enhance input preprocessing:

Filter or map unknown words to index 2

Add word validation during encoding step

Optional improvements:

Use Tokenizer and retrain with real-world data

Switch to LSTM/GRU for better generalization on longer, unseen reviews

